lib/seattleflu/db/cli/command/etl/presence_absence.py

```diff
@@ -0,0 +1,385 @@
+    if uuid:
+        #Check Identifier_Set
+        if uuid.name == "samples":
+            LOG.info(f"Found sample UUID {uuid.uuid}")
+            return uuid.uuid
+        else:
+            raise IncorrectIdentifierSetError(f"Found sample UUID {uuid.uuid}, but UUID is of the incorrect identifier set «{uuid.name}»")
+    else:
+        LOG.warning(f"No corresponding UUID found for barcode {barcode}")
+        return barcode
+
+
+def sample_details(document: dict) -> dict:
+    """
+    Describe sample details in a simple data structure designed to be used
+    from SQL.
+    """
+    return { 
+        "sample_comment": document['sampleComment'],
+        "initial_sequence_call": document['initialProceedToSequencingCall'],
+        "proceed_to_sequence": document["sampleProceedToSequencing"]
+    }
+
+def upsert_presence_absence(db: DatabaseSession,
+                            identifier: str,
+                            sample_id: int,
+                            target_id: int,
+                            present: bool,
+                            well_details: list) -> Any:
+    """
+    Upsert presence_absence by its *identifier*.
+    """
+    LOG.debug(f"Upserting presence_absence «{identifier}»")
+
+    data = {
+        "identifier": f"NWGC_{identifier}",
+        "sample_id": sample_id,
+        "target_id": target_id,
+        "present": present,
+        "well_details": Json(well_details)
+    }
+
+    presence_absence = db.fetch_row("""
+        insert into warehouse.presence_absence (
+                identifier,
+                sample_id,
+                target_id,
+                present,
+                well_details)
+            values (
+                %(identifier)s,
+                %(sample_id)s,
+                %(target_id)s,
+                %(present)s,
+                %(well_details)s)
+
+        on conflict (identifier) do update
+            set sample_id = excluded.sample_id,
+                target_id = excluded.target_id,
+                present   = excluded.present,
+                well_details = excluded.well_details
+
+        returning presence_absence_id as id, identifier
+        """, data)
+
+        # TODO: the safest thing to do in the above sql query, on conflict, is probably not 
+        # to look at the identifier for a conflict. For example, if the lab sending us the 
+        # data wants to re-run any tests, their backend may force their identifier id to change.
+        # Instead, we may want to set a constraint that any sample_id and target_id combination
+        # should be unique, and create a conflict rule in the situation where duplicates are
+        # uploaded. 
+
+    assert presence_absence.id, "Upsert affected no rows!"
+
+    LOG.info(f"Upserted presence_absence {presence_absence.id} «{presence_absence.identifier}»")
+
+    return presence_absence
+
+
+def get_target_result(target_status: str, sample_identifier: int, target_identifier: int) -> Any:
```

Comment:
Ditto re: passing in function args for logging purposes only. We should do the same thing here we decide upon for the control status function.

Reply:
Ditto re: `target_control` function decision.